x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -109,6 +109,55 @@ ResolvedIndices resolve(String action, TransportRequest request, Metadata metada
         return resolveIndicesAndAliases(action, (IndicesRequest) request, metadata, authorizedIndices);
     }
 
+    boolean requiresWildcardExpansion(IndicesRequest indicesRequest) {
+        if (indicesRequest instanceof IndicesAliasesRequest) {
+            return true;
+        } else {
+            return indicesRequest instanceof IndicesRequest.Replaceable;
+        }
+    }
+
+    ResolvedIndices resolveIndicesAndAliases(String action, IndicesRequest indicesRequest) {
+        assert false == requiresWildcardExpansion(indicesRequest) : "indices must not require wildcard expansion";
+        final String[] indices = indicesRequest.indices();
+        if (indices == null || indices.length == 0) {
+            throw new IllegalArgumentException("the action " + action + " requires explicit index names, but none were provided");
+        }
+        if (IndexNameExpressionResolver.isAllIndices(Arrays.asList(indices))) {
+            throw new IllegalArgumentException(
+                "the action "
+                    + action
+                    + " does not support accessing all indices;"
+                    + " the provided index expression ["
+                    + Strings.arrayToCommaDelimitedString(indices)
+                    + "] is not allowed"
+            );
+        }
+
+        // TODO: Shard level requests have wildcard expanded already and do not need go through this check
+        final List<String> wildcards = Stream.of(indices).filter(Regex::isSimpleMatchPattern).collect(Collectors.toList());
+        if (wildcards.isEmpty() == false) {
+            throw new IllegalArgumentException(
+                "the action "
+                    + action
+                    + " does not support wildcards;"
+                    + " the provided index expression(s) ["
+                    + Strings.collectionToCommaDelimitedString(wildcards)
+                    + "] are not allowed"
+            );
+        }
+
+        //NOTE: shard level requests do support wildcards (as they hold the original indices options) but don't support
+        // replacing their indices.
+        //That is fine though because they never contain wildcards, as they get replaced as part of the authorization of their
+        //corresponding parent request on the coordinating node. Hence wildcards don't need to get replaced nor exploded for
+        // shard level requests.
+        final List<String> localIndices = new ArrayList<>(indices.length);
+        for (String name : indices) {
+            localIndices.add(nameExpressionResolver.resolveDateMathExpression(name));
+        }
```

Comment:
Similarly, this loop is *not* needed if the request is not directly created for REST level request. That is, for request like ShardSearchRequest, we can skip it for some performance.